dashboard/app/models/shared_blockly_function.rb

```diff
@@ -12,17 +12,94 @@
 #  stack       :text(65535)
 #
 
+DEFINITION_BLOCK_TYPES = {
+  function: 'procedures_defnoreturn',
+  behavior: 'behavior_definition',
+}
+
+BLOCK_TYPES_BY_DEFINITION_TYPE = DEFINITION_BLOCK_TYPES.invert
+
 class SharedBlocklyFunction < ApplicationRecord
   enum block_type: {
     function: 0,
     behavior: 1,
   }
 
+  after_save :write_file
+  after_destroy :delete_file
   validates_presence_of :name
   validates_uniqueness_of :name
 
-  before_save do
-    self.block_type = 'behavior'
-    self.level_type = 'GameLabJr'
+  def file_xml
+    Nokogiri::XML::Builder.new do |xml|
+      xml.block(
+        type: DEFINITION_BLOCK_TYPES[block_type.to_sym],
+        deletable: false,
+        movable: false,
+        editable: false,
+      ) do
+        xml.mutation do
+          JSON.parse(arguments).each do |name, type|
+            xml.arg(name: name, type: type)
+          end
+          xml.description description
+        end
+        xml.title(name, name: 'NAME')
+        xml.statement(name: 'STACK') do
+          xml << stack
+        end
+      end
+    end.to_xml
+  end
+
+  def self.arguments_from_xml(args_xml)
+    JSON.generate(
+      args_xml.map do |arg_xml|
+        [arg_xml.attribute('name'), arg_xml.attribute('type')]
+      end.to_h
+    )
+  end
+
+  def xml_path(type=level_type, function_name=name)
+    Rails.root.join "config/shared_functions/#{type}/#{function_name}.xml"
```

Comment:
For blocks (where name must be globally unique) I think using `name` makes sense as the filename.  For functions, does it make sense to rely on something other than name?

`name` may still be the best option, just want to pose the question.

Reply:
I think it's still a good idea to enforce name uniqueness, even if we later decide to reference them some other way. I can't think of a situation where we really need two functions or behaviors with the same name.